File: packages/python/pezhwan/mfa.py
"""Multi-factor authentication (TOTP / backup codes)."""

from __future__ import annotations

from typing import Optional

from .client import PezhwanClient
from .models import Envelope, LoginTokens, MfaSetup, MfaChallenge


class Mfa:
    def __init__(self, client: PezhwanClient) -> None:
        self._client = client

    def setup(self) -> MfaSetup:
        envelope = self._client.post("/v1/mfa/setup", {})
        return MfaSetup(envelope.data or {})

    def enable(self, code: str) -> bool:
        envelope = self._client.post("/v1/mfa/enable", {"code": code})
        return bool((envelope.data or {}).get("enabled", envelope.success))

    def verify(self, code: str) -> bool:
        envelope = self._client.post("/v1/mfa/verify", {"code": code})
        return bool((envelope.data or {}).get("verified"))

    def disable(self, code: str) -> bool:
        envelope = self._client.post("/v1/mfa/disable", {"code": code})
        return envelope.success

    def login(self, user_id: str, code: str) -> tuple[LoginTokens, Optional[MfaChallenge]]:
        envelope = self._client.post("/v1/mfa/login", {"userId": user_id, "code": code}, auth=False)
        data = envelope.data or {}
        if data.get("mfaRequired"):
            return LoginTokens(), MfaChallenge(data)
        self._client.access_token = data.get("accessToken")
        return LoginTokens(data), None

File: packages/go/pezhwan/mfa.go
package pezhwan

import "context"

// Mfa implements the /v1/mfa* endpoints (TOTP + backup codes).
type Mfa struct {
	client *Client
}

// Setup starts TOTP enrollment and returns the secret + QR code.
func (m *Mfa) Setup(ctx context.Context) (MfaSetup, error) {
	data, err := m.client.post(ctx, "/v1/mfa/setup", map[string]any{})
	if err != nil {
		return MfaSetup{}, err
	}
	return MfaSetup{Secret: str(data["secret"]), QRCode: str(data["qrCode"])}, nil
}

// Enable activates MFA after verifying the first code.
func (m *Mfa) Enable(ctx context.Context, code string) error {
	_, err := m.client.post(ctx, "/v1/mfa/enable", map[string]string{"code": code})
	return err
}

// Verify performs a step-up verification of a current code.
func (m *Mfa) Verify(ctx context.Context, code string) (bool, error) {
	data, err := m.client.post(ctx, "/v1/mfa/verify", map[string]string{"code": code})
	if err != nil {
		return false, err
	}
	ok, _ := data["verified"].(bool)
	return ok, nil
}

// Disable turns MFA off (requires a valid current code).
func (m *Mfa) Disable(ctx context.Context, code string) error {
	_, err := m.client.post(ctx, "/v1/mfa/disable", map[string]string{"code": code})
	return err
}

// Login completes an MFA-challenged login and returns tokens.
func (m *Mfa) Login(ctx context.Context, userID, code string) (LoginTokens, error) {
	data, err := m.client.post(ctx, "/v1/mfa/login", map[string]string{"userId": userID, "code": code})
	if err != nil {
		return LoginTokens{}, err
	}
	if data != nil {
		if tok, ok := data["accessToken"].(string); ok && tok != "" {
			m.client.AccessToken = tok
		}
	}
	return tokensFrom(data), nil
}